src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -198,11 +201,48 @@ private function parseDefinition(\DOMElement $service, $file, array $defaults =
             } elseif (isset($defaults['public'])) {
                 $public = $defaults['public'];
             }
-            $this->container->setAlias((string) $service->getAttribute('id'), new Alias($alias, $public));
+            $this->container->setAlias($id, new Alias($alias, $public));
 
             return;
         }
 
+        $definition = $this->getDefinition($service, $file, $defaults);
+        if (null === $instanceof) {
+            return $definition;
+        }
+
+        $className = $definition->getClass() ?: $id;
+        $parentId = $definition instanceof ChildDefinition ? $definition->getParent() : null;
+        foreach ($instanceof as $s) {
+            $type = (string) $s->getAttribute('id');
+            if (!is_a($className, $type, true)) {
+                continue;
+            }
+
+            if ($parentId) {
+                $s->setAttribute('parent', $parentId);
+            }
+            // TODO: move the ID generation or maybe the whole block in the parent class
+            $parentId = md5("$file.$type.$id");
```

Comment:
Including file as is is a bad idea... It makes the md5 related to the absolute path of the file. It will create deployment problems and it's why test fails. I'll change that.